azure-iot-device/azure/iot/device/common/http_transport.py

```diff
@@ -0,0 +1,193 @@
+        else:
+            ssl_context.load_default_certs()
+        ssl_context.verify_mode = ssl.CERT_REQUIRED
+        ssl_context.check_hostname = True
+
+        if self._x509_cert is not None:
+            logger.debug("configuring SSL context with client-side certificate and key")
+            ssl_context.load_cert_chain(
+                self._x509_cert.certificate_file,
+                self._x509_cert.key_file,
+                self._x509_cert.pass_phrase,
+            )
+
+        return ssl_context
+
+    # def _format_headers(self, headers):
+    #     # TODO: Right now I'm only doing sas token because I'm lazy and limited
+    #     formatted_headers = {}
+    #     formatted_headers["Authorization"] = headers["sas"]
+    #     return formatted_headers
+
+    """
+    Some backstory on this.
+
+    1) During discussion the concept of multiple simultaneous requests being made was brought up.
+    Let's say you are a person and you want to make a request a bunch of times for some reason.
+    We could do two things down here. First option, we could have a single HTTPSConnection established
+    during instantation of the transport, in which case the HTTPSConnection would be shared among all
+    requests! Cool? Maybe. I don't know precisely, but I'm pretty sure that if the conneciton is making a
+    request, and then another thread comes along and using the same connection makes a request... well I just
+    don't really know what's going to happen. So the second (and currently implemented) option for this is to
+    create a new HTTPSConnection object every time we make a request, so that each request has it's own "connection".
+    What does this mean? Not sure, but it seem safer.
+
+    2) Initially I atomized each http_client call. We would create the client in one call, we would connect in another,
+    disconnect in yet another. Seemed fine... ish. But then Carter was like hey we should make all the http_client specific
+    stuff down here and only expose a relevant surface. What's a relevant surface for HTTP? Probably just request. So that's
+    what I'm exposing. Everything, the insantiation of the HTTPSConnection, the sending of the request, the getting of the response,
+    all happens in this one call, and that's better. Probably.
+    """
+
+    @pipeline_thread.invoke_on_http_thread_nowait
+    # TODO: This star syntax is incompatible with Python 2, change it so that the callback is in front of the optional params.
+    def request(self, method, hostname, path, callback, body=None, headers={}, query_params=None):
+        # Sends a complete request to the server
+        logger.info("sending https request")
+        try:
+            logger.debug("creating to https connection")
+            connection = http_client.HTTPSConnection(
+                self._hostname, context=self._create_ssl_context()
+            )
+            logger.debug("connecting to host tcp socket")
+            connection.connect()
```

Comment:
Why not do this when the transport is instantiated?

Reply:
yes this is an important discussion topic. The thing with HTTP is that every time it makes a request, it opens a connection. It's not like MQTT in this way, and it's also why I have no connection logic anywhere you'll notice. Because every time I make a request, I open a connection, aka connect to a socket on the remote server, then during the request the TLS stack actually opens a connection to the server for the POST, waits till a response is received, then closes the connection. 

The reason I don't have connect on instantiation is because each connection is specific to a request. It has to be this way if we want to allow multiple requests simultaneously.

---
In reply to: [351475326](https://github.com/Azure/azure-iot-sdk-python/pull/403#discussion_r351475326) [](ancestors = 351475326)